File: mzcd3-code/code/RecipesV3/PPRecipes/RecipeIngredientToIngredient.m
/***
 * Excerpted from "Core Data in Objective-C, Third Edition",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material,
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose.
 * Visit http://www.pragmaticprogrammer.com/titles/mzcd3 for more book information.
***/
#import "RecipeIngredientToIngredient.h"


@implementation RecipeIngredientToIngredient

- (BOOL)createDestinationInstancesForSourceInstance:(NSManagedObject*)src 
                                      entityMapping:(NSEntityMapping*)map 
                                            manager:(NSMigrationManager*)mgr
                                              error:(NSError**)error
{
  NSManagedObjectContext *destMOC = [mgr destinationContext];
  NSString *deName = [map destinationEntityName];

  NSString *name = [src valueForKey:@"name"];
  
  NSMutableDictionary *ui = (NSMutableDictionary*)[mgr userInfo];
  if (!ui) {
    ui = [NSMutableDictionary dictionary];
    [mgr setUserInfo:ui];
  }
  NSMutableDictionary *inLookup = [ui valueForKey:@"ingredients"];
  if (!inLookup) {
    inLookup = [NSMutableDictionary dictionary];
    [ui setValue:inLookup forKey:@"ingredients"];
  }
  
  NSMutableDictionary *uofmLookup = [ui valueForKey:@"unitOfMeasure"];
  if (!uofmLookup) {
    uofmLookup = [NSMutableDictionary dictionary];
    [ui setValue:uofmLookup forKey:@"unitOfMeasure"];
  }
  

  NSManagedObject *dest = [ingredientLookup valueForKey:name];
  if (!dest) {
    dest = [NSEntityDescription insertNewObjectForEntityForName:deName
                                         inManagedObjectContext:destMOC];
    [dest setValue:name forKey:@"name"];
    [inLookup setValue:dest forKey:name];
    
    name = [source valueForKey:@"unitOfMeasure"];
    NSManagedObject *uofm = [uofmLookup valueForKey:name];
    if (!uofm) {
      id entityName = @"UnitOfMeasure";
      uofm = [NSEntityDescription insertNewObjectForEntityForName:entityName
                                           inManagedObjectContext:destMOC];
      [uofm setValue:name forKey:@"name"];
      [dest setValue:uofm forKey:@"unitOfMeasure"];
      [uofmLookup setValue:uofm forKey:name];
    }
  }

  [manager associateSourceInstance:source
           withDestinationInstance:dest
                  forEntityMapping:mapping];
  return YES;
}

- (BOOL)createRelationshipsForDestinationInstance:(NSManagedObject*)dIn 
                                    entityMapping:(NSEntityMapping*)map 
                                          manager:(NSMigrationManager*)mgr 
                                            error:(NSError**)error
{
  return YES;
}

@end

File: mzswift-code/code/PPRecipesV3/PPRecipes/RecipeIngredientToIngredient.swift
//
//  RecipeIngredientToIngredient.swift
//  PPRecipes
//
//  Created by Marcus S. Zarra on 12/14/15.
//  Copyright © 2015 The Pragmatic Programmer. All rights reserved.
//

import Foundation
import CoreData

class RecipeIngredientToIngredient: NSEntityMigrationPolicy {

  override func createDestinationInstances(forSource s: NSManagedObject,
    in mapping: NSEntityMapping,
    manager: NSMigrationManager) throws {
    let destMOC = manager.destinationContext
    guard let dEntityName = mapping.destinationEntityName else {
      fatalError("Destination entity name is nil")
    }

    guard let name = s.value(forKey: "name") as? String else {
      fatalError("Source object did not have a value for 'name'")
    }

    var userInfo: [AnyHashable: Any]
    if let managerUserInfo = manager.userInfo {
      userInfo = managerUserInfo
    } else {
      userInfo = [AnyHashable: Any]()
    }

    var ingredientLookup: [String:NSManagedObject]!
    if let lookup = userInfo["ingredients"] as? [String:NSManagedObject] {
      ingredientLookup = lookup
    } else {
      ingredientLookup = [String:NSManagedObject]()
      userInfo["ingredients"] = ingredientLookup
    }

    var uofmLookup: [String:NSManagedObject]!
    if let lookup = userInfo["unitOfMeasure"] as? [String:NSManagedObject] {
      uofmLookup = lookup
    } else {
      uofmLookup = [String:NSManagedObject]()
      userInfo["unitOfMeasure"] = uofmLookup
    }

    var dest = ingredientLookup[name]
    if dest == nil {
      dest = NSEntityDescription.insertNewObject(forEntityName: dEntityName,
        into: destMOC)
      dest!.setValue(name, forKey:"name")
      ingredientLookup[name] = dest

      guard let uofmName = s.value(forKey: "unitOfMeasure") as? String else {
        fatalError("Unit of Measure name is nil")
      }
      var uofm = uofmLookup[uofmName]
      if uofm == nil {
        let eName = "UnitOfMeasure"
        uofm = NSEntityDescription.insertNewObject(forEntityName: eName,
          into: destMOC)
        uofm!.setValue(uofmName, forKey:"name")
        dest!.setValue(uofm, forKey:"unitOfMeasure")
        uofmLookup[name] = uofm
      }
    }
    manager.userInfo = userInfo
  }

  override func createRelationships(forDestination d: NSManagedObject,
    in mapping: NSEntityMapping,
    manager: NSMigrationManager) throws {
  }
}
